### ISSUE TITLE
ParcelAnnotationProcessor.process take lot amount of time.

### ISSUE THREAD

**Issue by doniwinata0309** [johncarl81/parceler#392]
Hi, from [asyncprofiler](https://github.com/gradle/gradle-profiler) i can see the ParcelAnnotationProcessor.process dominating my android build and cpu process. 

<img width="1262" alt="Screen Shot 2020-10-21 at 14 27 55" src="https://user-images.githubusercontent.com/38415463/96689369-82852a80-13ac-11eb-839f-0ec01d55725f.png">

it seems spend most of time to FileOutputStream.write.
<img width="1211" alt="Screen Shot 2020-10-21 at 14 43 22" src="https://user-images.githubusercontent.com/38415463/96688831-cd527280-13ab-11eb-837a-7ca8de448f16.png">

Do you have any clue what may causing this ? my project contains lot of parcel annotated class, so i am not sure if it expected or not for that process taking that long. 
Let me know if you need additional information. Thank you in advance.

I attached the flame chart below
[flamechart.zip](https://github.com/johncarl81/parceler/files/5414129/flamechart.zip)

**Comment by johncarl81** [comment]
This is interesting.  Could you give me a ballpark of how any parcel annotated classes you're working with?

**Comment by doniwinata0309** [comment]
Basically something like this: https://github.com/doniwinata0309/build-perf-test/blob/test/parceler/androidAppModule1/src/main/java/com/tes1.java

But some are more complex and extending from few parents, also contains field from custom object.

I just created 300+ dummy parcel class on this project
https://github.com/doniwinata0309/build-perf-test/tree/test/parceler 

That will produce this flamegraph

<img width="1244" alt="Screen Shot 2020-10-22 at 18 06 29" src="https://user-images.githubusercontent.com/38415463/96863627-598a9580-1491-11eb-833c-30f89271f362.png">

[android_build-6.7-flames.svg.zip](https://github.com/johncarl81/parceler/files/5422217/android_build-6.7-flames.svg.zip)

So it seems this process that taking so long
https://github.com/johncarl81/transfuse/blob/master/transfuse-support/src/main/java/org/androidtransfuse/transaction/CodeGenerationScopedTransactionWorker.java#L49

**Comment by johncarl81** [comment]
Right, ok... looking at it deeper, it looks like it's taking a long time to flush the FileOutputStream to disk.  Not sure if we'll be able to do anything about this.

A while ago we implemented incremental processing to the annotation processor - does it take this long every time, or only on the first clean build?

**Comment by doniwinata0309** [comment]
Our Incremental build doing pretty well,  we only got issue for clean/full build.

I see, it seems the huge number of input and disk IO speed that made it slower. Especially when there are several workers working in the same time. Normally 1 module may take 2 minute to complete, but when they are compiled in parallel (we set 5 workers can run in parallel) it will takes 4 minutes for each module to complete.  

Do you think we can do something with FileOutputStream  under JCodeModel library by modifying tranfuse library ? perhaps i can try some approach mentioned here: https://stackoverflow.com/a/20555164  or https://www.oracle.com/technical-resources/articles/javase/perftuning.html

**Comment by johncarl81** [comment]
Would buffering do the trick then? - maybe just wrap the given OutputStream [here](https://github.com/johncarl81/transfuse/blob/master/transfuse-core/src/main/java/org/androidtransfuse/gen/FilerSourceCodeWriter.java#L53) with `BufferedOutputStream` ?

**Comment by doniwinata0309** [comment]
possibly, i will try to modify it a little bit later. thank you for pointing that out.
If you also have some alternative please let me know, i will love to try it

**Comment by doniwinata0309** [comment]
I tried to modify it https://github.com/johncarl81/transfuse/blob/master/transfuse-core/src/main/java/org/androidtransfuse/gen/FilerSourceCodeWriter.java#L53
to this code:
  int bufferSize = 8 * 1024;
        OutputStream os = new BufferedOutputStream(
                resource.openOutputStream(),
                bufferSize
        );

But the issue not fixed. 

From this class:
https://github.com/johncarl81/transfuse/blob/master/transfuse-support/src/main/java/org/androidtransfuse/transaction/CodeGenerationScopedTransactionWorker.java#L49
But how FilerSourceCodeWriter used in that file ? It seems use CodeWriter instead FilerSourceCodeWriter right ?

**Comment by doniwinata0309** [comment]
ah sorry, it does use the FilerSourceCodeWriter, just realise it extending that class and i saw method printed in the flamegraph

**Comment by doniwinata0309** [comment]
<img width="1189" alt="Screen Shot 2020-10-27 at 18 08 20" src="https://user-images.githubusercontent.com/38415463/97294042-11e47f00-1880-11eb-8d87-0870b51468a2.png">
Seems to be the same even i use buffer with size 8kb and 32kb.

**Comment by johncarl81** [comment]
There may be no way around it.  I'll try to fire it up on my machine to see how slow it is on different hardware.

**Comment by doniwinata0309** [comment]
https://github.com/johncarl81/transfuse/issues/233

There is new issue to adress this, we can try to use JavaFileObject.openWriter as alternative. I will try it next week.

**Comment by doniwinata0309** [comment]
hi @johncarl81 
https://github.com/johncarl81/transfuse/pull/234
this PR  solve ~~my issue and my build is faster now~~ (sorry after few times rerun the scenario on CI and local, seems the build speed is pretty much the same with old parceler). However, the parceler annotation cpu usage now less than dagger and databinding.
![Screen Shot 2020-11-02 at 17 04 02](https://user-images.githubusercontent.com/38415463/97855621-98400b80-1d2d-11eb-8bd9-4a5f4d0c390c.png)



do you mind to check it later ? thanks

**Comment by johncarl81** [comment]
Can we close @doniwinata0309 ?

**Comment by doniwinata0309** [comment]
yes thank you. is it going to delivered on next release of parceler (1.1.14 i guess) ?

**Comment by johncarl81** [comment]
Yes
